Index: libjpeg-turbo-1.1.90+svn733/Makefile.am
===================================================================
--- libjpeg-turbo-1.1.90+svn733.orig/Makefile.am	2011-12-20 19:33:20.000000000 +0000
+++ libjpeg-turbo-1.1.90+svn733/Makefile.am	2011-12-20 19:35:26.358583864 +0000
@@ -72,8 +72,7 @@
 
 endif
 
-bin_PROGRAMS = cjpeg djpeg jpegtran rdjpgcom wrjpgcom tjbench
-noinst_PROGRAMS = tjunittest jcstest
+bin_PROGRAMS = cjpeg djpeg jpegtran rdjpgcom wrjpgcom tjbench tjunittest jcstest
 
 tjbench_SOURCES = tjbench.c bmp.h bmp.c tjutil.h tjutil.c rdbmp.c rdppm.c \
 	wrbmp.c wrppm.c
